pkg/trace/local_client.go

```diff
@@ -0,0 +1,219 @@
+		fileName := fmt.Sprintf("%s/%s.jsonl", path, table)
+		err := os.MkdirAll(path, 0700)
+		file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open or create file %s: %w", fileName, err)
+		}
+		fm[table] = file
+	}
+
+	lc := &LocalClient{
+		fileMap: fm,
+		cfg:     cfg,
+		canal:   make(chan Event[Entry], cfg.Instrumentation.TraceBufferSize),
+		chainID: chainID,
+		nodeID:  nodeID,
+		logger:  logger,
+	}
+
+	go lc.draincanal()
+
+	return lc, nil
+}
+
+func (lc *LocalClient) Write(e Entry) {
+	if !lc.IsCollecting(e.Table()) {
+		return
+	}
+	lc.canal <- NewEvent(lc.chainID, lc.nodeID, e.Table(), e)
+}
+
+// ReadTable returns a file for the given table. If the table is not being
+// collected, an error is returned. This method is not thread-safe.
+func (lc *LocalClient) ReadTable(table string) (*os.File, error) {
+	file, has := lc.getFile(table)
+	if !has {
+		return nil, fmt.Errorf("table %s not found", table)
+	}
+
+	return file, nil
+}
+
+func (lc *LocalClient) IsCollecting(table string) bool {
+	if _, has := lc.getFile(table); has {
+		return true
+	}
+	return false
+}
+
+// getFile gets a file for the given type. This method is purposely
+// not thread-safe to avoid the overhead of locking with each event save.
+func (lc *LocalClient) getFile(table string) (*os.File, bool) {
+	f, has := lc.fileMap[table]
+	return f, has
+}
+
+// saveEventToFile marshals an Event into JSON and appends it to a file named after the event's Type.
+func (lc *LocalClient) saveEventToFile(event Event[Entry]) error {
+	file, has := lc.getFile(event.Table)
+	if !has {
+		return fmt.Errorf("table %s not found", event.Table)
+	}
+
+	eventJSON, err := json.Marshal(event)
+	if err != nil {
+		return fmt.Errorf("failed to marshal event: %v", err)
+	}
+
+	if _, err := file.Write(append(eventJSON, '\n')); err != nil {
+		return fmt.Errorf("failed to write event to file: %v", err)
+	}
+
+	return nil
+}
+
+// draincanal takes a variadic number of channels of Event pointers and drains them into files.
+func (lc *LocalClient) draincanal() {
+	// purposefully do not lock, and rely on the channel to provide sync
+	// actions, to avoid overhead of locking with each event save.
+	for ev := range lc.canal {
+		if err := lc.saveEventToFile(ev); err != nil {
+			lc.logger.Error("failed to save event to file", "error", err)
+		}
+	}
+}
+
+// Stop optionally uploads and closes all open files.
+func (lc *LocalClient) Stop() {
+	for table, file := range lc.fileMap {
+		if lc.cfg.Instrumentation.TracePushURL != "" {
+			err := UploadFile(lc.cfg.Instrumentation.TracePushURL, lc.chainID, lc.nodeID, table, file)
+			if err != nil {
+				lc.logger.Error("failed to upload trace file", "error", err)
+			}
+		}
+		file.Close()
+	}
+}
+
+// UploadFile uploads a file to the given URL. This function does not close the
+// file.
+func UploadFile(url, chainID, nodeID, table string, file *os.File) error {
```

Comment:
got uploading to work I think, but didn't get the server to actually receive it for push or pulling